File: frontend/src/app/services/show.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Globals} from '../global/globals';
import {Show} from '../dtos/show';
import {ShowSeat} from '../dtos/show-seat';
import {Price} from '../dtos/price';
import {SaveShow} from '../dtos/saveShow';
import {OrderTickets} from '../dtos/orderTickets';
import {Ticket} from '../dtos/ticket';

@Injectable({
  providedIn: 'root'
})
export class ShowService {

  private showBaseUri: string = this.globals.backendUri + '/shows';


  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Loads specific show from the backend
   * @param id of show to load
   */
  getShowById(id: number): Observable<Show> {
    console.log('Load show details for ' + id);
    return this.httpClient.get<Show>(this.showBaseUri + '/' + id);
  }

  /**
   * Loads a list of seats with their status for a specific show from the backend
   * @param id of show
   */
  getSeatsOfShow(id: number): Observable<ShowSeat[]> {
    console.log('Load seats of show for ' + id);
    return this.httpClient.get<ShowSeat[]>(this.showBaseUri + '/' + id + '/seats');
  }

  /**
   * Loads a list of prices for a specific show from the backend
   * @param id of the show
   */
  getPricesOfShow(id: number): Observable<Price[]> {
    console.log('Load prices of show for ' + id);
    return this.httpClient.get<Price[]>(this.showBaseUri + '/' + id + '/prices');
  }

  /**
   * Persists a show (and its tickets) to the backend
   * @param saveShow to persist
   */
  createShow(saveShow: SaveShow): Observable<Show> {
    console.log('Create show with title ' + saveShow.show.title);
    return this.httpClient.post<Show>(this.showBaseUri, saveShow);
  }

  /**
   * Loads all shows as page
   * @param page to be shown
   */
  getAllShows(page: number): Observable<Show[]> {
    console.log('Get page nr. ' + page + ' of shows');
    return this.httpClient.get<Show[]>(this.showBaseUri + '/pages/' + page);
  }

  /**
   * Order tickets for a show
   */
  orderTickets(saveTickets: OrderTickets, showId: number): Observable<number[]> {
    console.log('Order tickets for show ' + showId + ' [' + saveTickets.mode + ']');
    return this.httpClient.post<number[]>(this.showBaseUri + '/' + showId + '/order', saveTickets);
  }

  /**
   * Purchase previously reserved tickets
   */
  purchaseReservedTickets(ticketIds: number[]): Observable<number> {
    console.log('Purchase previously reserved tickets ' + ticketIds);
    return this.httpClient.post<number>(this.showBaseUri + '/purchase-reserved', ticketIds);
  }

  exportPurchasedTicketPdf(orderId: number): Observable<Blob> {
    return this.httpClient.get(this.showBaseUri + '/generatepdf/purchase/' + orderId, {responseType: 'blob'});

  }

  exportCancelTicketPdf(orderId: number, ticketIds:number[], bonusPointsInEuro:number): Observable<Blob> {
      return this.httpClient.post(this.showBaseUri + '/generatepdf/cancel/' + orderId + '/' + bonusPointsInEuro , ticketIds , {responseType: 'blob'});

    }


  cancelTicket(ticketIds: number[]): Observable<number>  {
      console.log('Cancel tickets with order Ids: ', ticketIds);
      return this.httpClient.post<number>(this.showBaseUri + '/cancel-tickets' , ticketIds);
  }

  /**
   * Loads all ticket if specified user
   * @param userId id of specified user
   */
  getAllTicketsByUserId() {
    console.log('Sending get request for all ticket');
    return this.httpClient.get<Ticket[]>(this.showBaseUri + '/my-orders');
  }

  getTicketPagination(page: number) {
    console.log('Load all tickets page: ' + page);
    const params = new HttpParams().set('page', String(page));
    return this.httpClient.get(this.showBaseUri + '/tickets-pagination', {params});
  }

  /**
   * Loads all shows from BE withe pages
   */
  getShowPagination(page: number) {
    console.log('Load all show');
    const params = new HttpParams().set('page', String(page));
    return this.httpClient.get(this.showBaseUri + '/pagination', {params});
  }

  /**
   * searching shows for title
   * @param name of show which is searched
   * @param page page for pagination
   */
  searchShow(title: string, location: string, price: number, date: string, time:string, page: number) {
    console.log('searching shows with this title ' + title + page + location + price);
    const params = new HttpParams().set('title', title).set('location', location).set('price', String(price)).set('date', date).set('time', time).set('page', String(page));
    return this.httpClient.get(this.showBaseUri + '/search', {params});
  }

  /**
   * get all the events from an artist
   * @param id of artist
   */
  getShowsByLocation(id: number, page: number): Observable<Show[]> {
    console.log('get shows for location with this id:' + id);
    const params = new HttpParams().set('id', String(id)).set('page', String(page));
    return this.httpClient.get<Show[]>(this.showBaseUri + '/location', {params});
  }

}

File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/ShowService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.SaveShow;
import at.ac.tuwien.sepm.groupphase.backend.entity.Seat;
import at.ac.tuwien.sepm.groupphase.backend.entity.Show;
import at.ac.tuwien.sepm.groupphase.backend.entity.Ticket;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException;
import org.springframework.core.io.InputStreamResource;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;

import java.util.List;
import java.util.Map;

public interface ShowService {
    /**
     * Find all show entries.
     *
     * @return list of all show entries
     */
    List<Show> findAll();

    /**
     * Find all show entries as page.
     *
     * @param pageable for  pagination
     * @return Page of show entries
     */
    Page<Show> findAllAsPage(Pageable pageable);

    /**
     * Get all tickets of a show by id.
     *
     * @param id of show
     * @return list of all tickets of the specified show
     * @throws NotFoundException if there is no show with this id
     */
    List<Ticket> getTicketsOfShow(Long id);

    /**
     * Get all seats of a show by id.
     *
     * @param id of show
     * @return list of all seats present at the location where the specified show is held
     * @throws NotFoundException if there is no show with this id
     */
    List<Seat> getSeatsOfShow(Long id);

    /**
     * Get all seats with their corresponding ticket (or null if seat is "NOSEAT").
     *
     * @param id of ticket
     * @return map of seats and tickets, every seat is mapped to its ticket (or to null if seat is "NOSEAT")
     * @throws NotFoundException if there is no show with this id
     */
    Map<Seat, Ticket> getSeatsWithTicket(Long id);

    /**
     * Find a single show entry by id.
     *
     * @param id od show
     * @return the show entry
     * @throws NotFoundException if there is no show with this id
     */
    Show findOne(Long id);


    /**
     * Save a new show.
     *
     * @param saveShow show to save
     * @return saved show
     */
    Show saveShow(SaveShow saveShow);

    /**
     * Order tickets and updates the users bonus points according the sum of the purchased tickets.
     *
     * @param seatIds   of seats to order
     * @param showId    of show to order tickets from
     * @param userEmail of user who wants to order tickets
     * @param mode      either "RESERVE" or "PURCHASE"
     * @return orderId and the bonus point credit for this purchase.
     */
    Long[] orderTickets(List<Long> seatIds, Long showId, String userEmail, String mode);

    /**
     * Purchase previously reserved tickets.
     * Reserved tickets of the same order but not selected for purchase will be set to FREE.
     *
     * @param ticketIds of tickets to purchase
     * @param userEmail of user who wants to purchase tickets
     * @return order id
     */
    Long purchaseReservedTickets(List<Long> ticketIds, String userEmail);

    /**
     * Get tickets for a user by order id.
     *
     * @param userEmail of user
     * @param orderId   of tickets
     * @return tickets in that order
     * @throws NotFoundException if there is no order with this id
     */
    List<Ticket> getTicketsForUserByOrderId(Long orderId, String userEmail);

    /**
     * Cancel a ticket with given id.
     *
     * @param ticketsId list of ticket ids to be deleted
     * @param userEmail current user
     * @return the number of bonus points the user is charged, because hhe/she does not have enough.
     * @throws NotFoundException if ticket is not found
     */
    long cancelTickets(List<Long> ticketsId, String userEmail) throws IllegalAccessError;


    /**
     * Load all tickets of the user.
     *
     * @param id id of the user
     * @return List of Ticket entities
     */
    List<Ticket> getAllTicketsByUserId(Long id);

    /**
     * Find all show entries ordered by name.
     *
     * @param page for pagination
     * @return ordered list of all show entries
     */
    Page<Show> findAllByOrderByTitle(int page);

    /**
     * Find shows with the given name.
     *
     * @param title    of show to be found
     * @param location of show to be found
     * @param page     for pagination
     * @param price    for searched show
     * @param date     of searched dhow
     * @param time     of searched show
     * @return a list of artist
     */
    Page<Show> findByParams(String title, String location, int price, String date, String time, Integer page);


    /**
     * find all shows for the location with id.
     *
     * @param id       of location
     * @param pageable for pagination
     * @return a List of Shows from location with id
     */
    Page<Show> findShowsByLocation(Long id, Pageable pageable);

    Page<Ticket> getPageWithAllTicketsByUserId(Long id, int page);

    /**
     * generated PDF bill for canceled tickets.
     *
     * @param orderId     id of current order
     * @param ids         of tickets that are deleted to be displayed on the bill
     * @param email       from current user
     * @param bonusPoints add bonusPoints to tickets
     * @return PDF bill with canceled tickets
     */

    ResponseEntity<InputStreamResource> ticketCancelReport(Long orderId,
                                                           List<Long> ids,
                                                           String email,
                                                           Long bonusPoints);

    /**
     * generate PDF for purchased tickets.
     *
     * @param orderId id of current order
     * @param email   from current user
     * @return PDF bill with canceled tickets
     */
    ResponseEntity<InputStreamResource> ticketPurchaseReport(Long orderId, String email) throws Exception;

}
